File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.java
import java.io.BufferedReader;
import java.io.IOError;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    public static class Node{
        int x;
        int time;
        public Node(int x, int time){
            this.x=x;
            this.time=time;
        }
    }
    public static int bfs(int start_num, int end_num) {
        int min = Integer.MAX_VALUE;
        int max_size = 100001;
        int[] visited = new int[max_size];
        Queue<Node> queue = new LinkedList<>();
        visited[start_num] = 1;
        queue.offer(new Node(start_num,0));

        while (!queue.isEmpty()) {
            Node current  = queue.poll();
            if(current.x == end_num){
                min=Math.min(min,current.time);
            }

            if(current.x*2<max_size &&visited[current.x*2]!=1){
                visited[current.x*2]=1;
                queue.offer(new Node(current.x*2,current.time));
            }
            if(current.x-1>=0&&visited[current.x-1]!=1){
                visited[current.x-1]=1;
                queue.offer(new Node(current.x-1, current.time+1));
            }
            if(current.x+1<max_size &&visited[current.x+1]!=1){
                visited[current.x+1]=1;
                queue.offer(new Node(current.x+1, current.time+1));
            }
        }
        return min;

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String input= br.readLine();
        StringTokenizer st = new StringTokenizer(input);
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        System.out.println(bfs(n,k));


    }
}

File: 백준/Gold/13549. 숨바꼭질 3/숨바꼭질 3.py
from collections import deque
#bfs
arr=[2,-1,1]
def bfs (start, end):
    MAX=100001
    visited=[False]*MAX
    time = [0] * MAX

    queue= deque([start])
    visited[start] = True
    
    while queue:
        node=queue.popleft()

        if node==end:
            return time[node]

        for i in range(3):
            if i==0:
                next=node*arr[i]
                if 0<=next<MAX and not visited[next]:
                    
                    visited[next]=True
                    time[next]=time[node]
                    queue.appendleft(next)
                                
            else:
                next=node+arr[i]
                if 0<=next<MAX and not visited[next]:
                    
                    visited[next]=True
                    time[next]=time[node]+1
                    queue.append(next)
    return 0
    

    



n,k=map(int, input().split())
print(bfs(n,k))